test/puppetlabs/http/client/metrics_test.clj

```diff
@@ -0,0 +1,481 @@
+                    bytes-read-id-with-verb (str base-metric-id ".GET" ".bytes-read")]
+                (is (= (set (list bytes-read-id bytes-read-id-with-verb))
+                       (set (keys client-metrics))
+                       (set (keys client-metrics-data))))
+                (is (every? #(instance? Timer %) (vals client-metrics)))
+                (let [bytes-read-data (get client-metrics-data bytes-read-id)]
+                  (is (every? #(instance? ClientMetricData %) (vals client-metrics-data)))
+
+                  (is (= 1 (.getCount bytes-read-data)))
+                  (is (= bytes-read-id (.getMetricId bytes-read-data)))
+                  (is (<= 1000 (.getMean bytes-read-data)))
+                  (is (<= 1000 (.getAggregate bytes-read-data))))))))))
+     (testing "metrics work for failed request"
+       (try
+         (testwebserver/with-test-webserver-and-config
+          (unbuffered-test/blocking-handler data) port {:shutdown-timeout-seconds 1}
+          (let [metric-registry (MetricRegistry.)]
+            (with-open [client (-> (ClientOptions.)
+                                   (.setSocketTimeoutMilliseconds 200)
+                                   (.setConnectTimeoutMilliseconds 100)
+                                   (Async/createClient metric-registry))]
+              (let [request-options (doto (RequestOptions. (str "http://localhost:" port "/hello"))
+                                      (.setAs ResponseBodyType/UNBUFFERED_STREAM))
+                    response (-> client (.get request-options) .deref)
+                    error (.getError response)
+                    body (.getBody response)]
+                (is (nil? error))
+                (is (thrown? SocketTimeoutException (slurp body)))
+                (let [client-metrics (.getClientMetrics client)
+                      client-metrics-data (.getClientMetricsData client)
+                      base-metric-id (str "puppetlabs.http-client.experimental.with-url.http://localhost:" port "/hello")
+                      bytes-read-id (str base-metric-id ".bytes-read")
+                      bytes-read-id-with-verb (str base-metric-id ".GET" ".bytes-read")]
+                  (is (= (set (list bytes-read-id bytes-read-id-with-verb))
+                         (set (keys client-metrics))
+                         (set (keys client-metrics-data))))
+                  (is (every? #(instance? Timer %) (vals client-metrics)))
+                  (let [bytes-read-data (get client-metrics-data bytes-read-id)]
+                    (is (every? #(instance? ClientMetricData %) (vals client-metrics-data)))
+
+                    (is (= 1 (.getCount bytes-read-data)))
+                    (is (= bytes-read-id (.getMetricId bytes-read-data)))
+                    (is (<= 1 (.getMean bytes-read-data)))
+                    (is (<= 1 (.getAggregate bytes-read-data)))))))))
+         (catch TimeoutException e
```

Comment:
These tests are wrapped in `try/catch`'s because the webserver handling the client requests is being shutdown before it has completed handling these requests. Jetty throws a `TimeoutException` if its `stopTimeout` is reached before it gracefully completes all open requests. I've got a branch upgrading to tk-w-j9 1.5.4 (where we changed this behavior for TK-338) and removing these `try/catch`'s, I just wasn't sure it was a great idea to add that upgrade to a PR that's been open for a while and has a lot of commits.